### PR TITLE
O(1) rebalance

### PR DESCRIPTION
**Pull request by crusaderky**
- DONE implementation
- DONE stress test
- DONE unit tests

# In scope
- Make rebalance() O(1) to the total number of tasks on the cluster.
- Introduce hysteresis and thresholds to minimise the overall cost of rebalancing.
- Consider non managed memory (leaks, fragmentation, etc.) while rebalancing.
- Change algorithm that picks which tasks to move from largest first to least recently inserted first.

# Out of scope, left for future PRs
- This PR does not make rebalance safe to run while computations are running (nothing changes on that front).
- This PR does not harmonize / deduplicate the code that moves data around used by rebalance(), replicate(), and retire_workers().
- High level documentation containing recommendations regarding the use of malloc_trim, jemalloc, etc.

# Demo
https://gist.github.com/crusaderky/c1ccf5fd0107b13c8d24bbed5197d5f6
(also works on master)

### CODE REVIEW

**Comment by crusaderky** [comment]
CC @jrbourbeau @mrocklin

**Comment by crusaderky** [comment]
rebalance() runtime when there is nothing to do:

| # keys on cluster | # workers | master | this PR |
| --- | --- | --- | --- |
| 10k | 32 | 18.3 ~ 25.3 ms | 1.1 ~ 1.9 ms |
| 100k | 32 | 232 ~ 260 ms | 1.0 ~ 2.6 ms |

**Comment by crusaderky** [comment]
### Important
I noticed that, once the size of a single minimal Python object (read: the buffer of a numpy array) drops below 2 MiB, CPython / Linux x64 won't release RAM instantly anymore when the object is deallocated, but instead hog it indefinitely. gc.collect() does nothing. malloc_trim() typically, but not always, fixes the problem. Creating new Python objects simply reuses said memory. This is the behaviour already in master - nothing changes there - and it already impacts all logic that relies on measuring process memory (spilling, pausing, and restarting). What does change with this PR is that now that unused, allocated memory is considered by rebalance too(), which in turn means that keys may be evicted too aggressively from particularly heavy nodes, as there is currently no way to tell apart trimmable memory from memory leaks or genuine fragmentation.

The current workaround is that I let the user pick what measure he wants to use in rebalance(); by default is the optimistic memory (managed by dask in RAM + unmanaged older than 30s) but it can be changed to managed only, thus reverting to the behaviour of master. Needless to say this is not great as it adds config burden on the shoulders of the user.

An alternative, cleaner solution would be to run malloc_trim() periodically (e.g. 2~5s) on the workers, but problems will ensue if someone compiled the Python interpreter with an alternative alloc/free pair of primitives.

I have not tested the behaviour on MacOSX and Windows yet, but from what I saw in my previous PR I expect similar headaches there too.

File: distributed/distributed-schema.yaml [hunk 1]
```diff
@@ -404,6 +404,49 @@ properties:
...
+                  measure:
+                    enum:
+                      - process
+                      - optimistic
+                      - managed
+                      - managed_in_memory
+                    description: >-
+                      Which of the properties of distributed.scheduler.MemoryState
+                      should be used for measuring worker memory usage
+                  sender_min:
```
Comment (mrocklin) [inline distributed/distributed-schema.yaml:421]: Small note: most configuration names today use hyphens by default.  I think that get/set are intentionally blind to this difference, but we may want to stay consistent

Comment (crusaderky) [reply distributed/distributed-schema.yaml:421]: ok will change

**Comment by mrocklin** [comment]
> An alternative, cleaner solution would be to run malloc_trim() periodically (e.g. 2~5s) on the workers, but problems will ensue if someone compiled the Python interpreter with an alternative alloc/free pair of primitives.

Can I ask you to raise an issue about this and tag `@pitrou` and `@jakirkham`.  I don't know much about this, but it seems like this might be helpful in the common case.  "leaking" memory is a very common pain point today.

**Comment by fjetter** [comment]
> An alternative, cleaner solution would be to run malloc_trim() periodically (e.g. 2~5s) on the workers, but problems will ensue if someone compiled the Python interpreter with an alternative alloc/free pair of primitives.

cc @xhochy

**Comment by xhochy** [comment]
> > An alternative, cleaner solution would be to run malloc_trim() periodically (e.g. 2~5s) on the workers, but problems will ensue if someone compiled the Python interpreter with an alternative alloc/free pair of primitives.
> 
> cc @xhochy

Thanks for the ping, I'll probably should comment on the (to-be-opened) issue instead of here?

File: distributed/scheduler.py [hunk 2]
```diff
@@ -566,7 +584,7 @@ def __init__(
         )
 
         self._actors = set()
-        self._has_what = set()
+        self._has_what = {}
```
Comment (fjetter) [inline distributed/scheduler.py:587]: This feels a bit fragile and couples things I would rather have independent.

This assumption is used in the algorithm below when defining `iter(ws._has_what)` as the tasks which are potential victims for the rebalancing. (Note, that this does _not_ use the indirection via the property `has_what`)
If we are shooting for a last inserted, I believe I would prefer to make this explicit and pay for it, e.g. `sorted(ws._has_what, key=lambda ws: ws.last_update)` (whatever last_update would be). This would also allow for an easier modification of this condition.
However, it works and it is cheap. If my proposal is too expensive I can probably live with this.

**Review by fjetter — COMMENTED** [review]
I like the simplicity and elegance of the algorithm. In particular that there are no opaque heuristics in there. Other than configuring the different thresholds, I think we have a nice lever to control different _policies_ via the heap sort key (or rather the who_has sorting / insertion order / etc.)

I am a bit concerned about the dynamic case where this decision might become much more complicated. I would like to avoid the complexity we currently have in `dask.order`.
Also, I'm wondering how we can manage "data ownership", i.e. is a worker actually supposed to hold a replica or is it just a temp copy needed for a dependency, etc. I believe we do not have a data model for this, yet.

These concerns should not stop us here but rather should be kept in mind once we move on to the next step(s).

File: distributed/scheduler.py [hunk 3]
```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
...
+        # farthest from the cluster mean are at the top of the heaps.
+        # Heap elements:
+        # - maximum number of bytes to send or receive
+        # - number of bytes after which the worker should not be considered anymore
+        # - arbitrary unique number, there just to to make sure that WorkerState objects
+        #   are never used for sorting in the unlikely event that two processes have
+        #   exactly the same number of bytes allocated.
+        # - WorkerState
+        # - iterator of all tasks in memory on the worker (senders only)
+        senders: "list[tuple[int, int, int, WorkerState, Iterator[TaskState]]]" = []
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5632]: it's a python 3.9 style annotation. mypy can handle it on python 3.7+ as long as you have "from __future__ import annotations" in your header, which however Cython does not support.

File: distributed/scheduler.py [hunk 4]
```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
...
+        # Output: [(sender, recipient, task), ...]
+        msgs: "list[tuple[WorkerState, WorkerState, TaskState]]" = []
+
+        # optimistic memory = RSS - unmanaged memory that appeared over the last 30
+        # seconds (distributed.worker.memory.recent_to_old_time).
+        # This lets us ignore temporary spikes caused by task heap usage.
+        memory_by_worker = [
+            (ws, getattr(ws.memory, MEMORY_REBALANCE_MEASURE)) for ws in workers
+        ]
+        mean_memory = sum(m for _, m in memory_by_worker) // len(memory_by_worker)
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5644]: I need to store the memory measure somewhere for later (calling WorkerState.memory is mildly expensive). List comprehension tends to be a lot faster than explicit for loop + appends.

File: distributed/scheduler.py [hunk 5]
```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
...
-                    workers_by_task = {ts: ts._who_has for ts in tasks}
+            if keys is not None:
+                if not isinstance(keys, Set):
+                    keys = set(keys)  # unless already a set-like
+                if not keys:
+                    return {"status": "OK"}
+                tasks = [self._tasks[k] for k in keys]
+                missing_data = [ts._key for ts in tasks if not ts._who_has]
+                if missing_data:
+                    return {"status": "missing-data", "keys": missing_data}
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5578]: This is exclusively for the (edge?) case where a user explicitly states keys, and it respects the previous design. I am not familiar with who would explicitly pass the keys or workers arguments or their use cases, and I saw no reason to break their user interface.

File: distributed/scheduler.py [hunk 6]
```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
...
+
+            if (
+                ws._has_what
+                and ws_memory >= mean_memory + half_gap
+                and ws_memory >= sender_min
+            ):
+                # This may send the worker below sender_min (by design)
+                nbytes = mean_memory - ws_memory  # negative
+                senders.append(
+                    (nbytes, nbytes + half_gap, id(ws), ws, iter(ws._has_what))
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5664]: it's already documented above (see definition of senders) but I will add a reminder here.
You need to store an iter object because you want to go through all tasks of each sender from oldest to newest, but you want to pick tasks from senders that need to shed off more first in case there's not enough capacity in the receivers.

We're not abusing anything - the lifetime of the iter is contained within a non-async function in a single-threaded process.

File: distributed/scheduler.py [hunk 7]
```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
...
+                if nbytes + snd_nbytes_max > 0:
+                    # Moving this task would cause the sender to go below mean
+                    # and potentially risk becoming a recipient, which would
+                    # cause tasks to bounce around. Move on to the next task
+                    # of the same sender.
+                    continue
+
+                # Find the recipient, farthest from the mean, which
+                # 1. has enough available RAM for this task
+                # 2. don't hold a copy already
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5704]: This PR preserves the current levels of data duplication. It is not designed to run during computation (that's the target for a later PR) and reserves no special treatment to keys that have been temporarily duplicated for the purpose of a computation.

File: distributed/scheduler.py (hunk 2 above)
Comment (crusaderky) [reply distributed/scheduler.py:587]: the key point of this PR is to reduce the big-O complexity towards the total number of keys on the cluster from O(n*log(n)) to O(1). ``sorted`` would bring it back to O(n*log(n)) .

File: distributed/scheduler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:5664]: With abusing I rather mean: Are relying on this iterator to be initialized here for the algorithm to finish or is the code section below self contained and will terminate regardlessly

File: distributed/scheduler.py (hunk 2 above)
Comment (fjetter) [reply distributed/scheduler.py:587]: > O(n*log(n)) to O(1). sorted would bring it back to O(n*log(n)) .

What is N in this equation? My suggestion would indeed put the (worst case) complexity to N logN where N is the number of keys on a given worker, not the total number of keys on the cluster. While formally all keys could live on a single worker, this is an edge case which is highly unlikely and should imho not drive this implementation

If we ever want to move away from the "last inserted" policy we will need to touch this again


My actual issue is not necessarily the complexity or extensibility but I have rather that we rely on a dict to achieve this _last inserted_ policy. The data structure which comes to my mind for something like this would be a stack. In python, that's most easily achieved using a `deque`.
Relying on how the dict is sorted is an implementation detail of CPython. Afaik, they guarantee that dicts stay sorted since py3.6(?) but afaik they do not guarantee _how_ they keys are sorted.
Not only do we rely on it to stay this way, I feel that this knowledge makes the code a little harder to maintain.

Comment (crusaderky) [reply distributed/scheduler.py:587]: N is the total number of keys on the cluster. The docstring gets thoroughly into the details of the big-O equation.

> The data structure which comes to my mind for something like this would be a stack. In python, that's most easily achieved using a deque.

keys must be removed from the middle of the data structure under normal operations. A deque does not offer this.

> Relying on how the dict is sorted is an implementation detail of CPython. 

It is not. dicts became insertion-sorted as an implementation detail in Python 3.6, which was then elevated to formal API spec in Python 3.7. All alternative implementations must implement the same to be called Python.

> but afaik they do not guarantee how they keys are sorted.

Yes there is. dicts are sorted by insertion order.

Please refer to https://docs.python.org/3.7/whatsnew/3.7.html

> the insertion-order preservation nature of dict objects has been declared to be an official part of the Python language spec.

And to https://docs.python.org/3/reference/datamodel.html

> Dictionaries preserve insertion order, meaning that keys will be produced in the same order they were added sequentially over the dictionary. Replacing an existing key does not change the order, however removing a key and re-inserting it will add it to the end instead of keeping its old place.

>  I feel that this knowledge makes the code a little harder to maintain.

I am extensively documenting that we are relying on this feature, and I will add thorough unit tests to this same PR to prevent accidental regressions.

File: distributed/scheduler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/scheduler.py:5664]: Does the code rely on this iterator to be exhausted? IIUC the iterator is used in multiple loop iterations and the termination condition depends on this iterator to eventually be empty.

Put differently: Would the below algorithm work if this was a `list(ws._has_what)` instead of an `iter`?

[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5664]: > Does the code rely on this iterator to be exhausted?

Iterators will typically *not* be exhausted. The algorithm continues iterating on the keys of a worker until

a) all keys are exhausted (by having been moved to other workers, or by having been skipped because no suitable receivers were found), OR
b) the worker has dropped below the threshold.

Use case (a) exclusively happens when the unmanaged memory + keys that cannot be moved on their own are sufficient to go above the threshold.

> Would the below algorithm work if this was a list(ws._has_what) instead of an iter?

if you mean creating a list of the keys and then popping one element from the left of the list until empty, yes it would have been the same just marginally slower and less readable.

File: distributed/scheduler.py [hunk 8]
```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
...
-                else:
-                    workers = set(parent._workers_dv.values())
-                    workers_by_task = {ts: ts._who_has for ts in tasks}
+            if keys is not None:
+                if not isinstance(keys, Set):
+                    keys = set(keys)  # unless already a set-like
+                if not keys:
+                    return {"status": "OK"}
+                tasks = [self._tasks[k] for k in keys]
+                missing_data = [ts._key for ts in tasks if not ts._who_has]
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5576]: fixed

File: distributed/scheduler.py (hunk 2 above)
Comment (crusaderky) [reply distributed/scheduler.py:589]: > try to decouple normal operations from rebalancing.  That wouldn't be the first time we're actually tracking data twice. 

The alternative to using a dict is to write a set-like class that is internally implemented with a dict. But I found that would be overkill and actually make everything harder to read.

Comment (fjetter) [reply distributed/scheduler.py:589]: > > At this point we're not dealing with all tasks but only tasks in memory on a worker.
>
> Wrong, you're dealing with all tasks on all workers that need to donate keys

How is that wrong? The `has_what` is literally defined as the subset of tasks in memory. There is no general way to relate this number to the total number of tasks. There may be graphs with millions of tasks but only one final result key in memory. Only when scattering data, there is a 1to1 relation but for all others, this is non trivial.

---

I do not necessarily want to argue this academically. I do get your point about how the numbers can be related to the total number and that results in a big O of N log N but I argue that is only a valid argument given certain assumptions. these assumptions are sometimes true and sometimes not. If they are true, the question is whether this is relevant or not and how this impacts reality is difficult to argue about on a theoretical level here. At least it would require a more structured approach than what we can do in a GH thread, imho.

The bigger point I am worried about is that **if** this insertion-sorted `has_what` is what makes or breaks the rebalancing and we do not even see any way to work around this or evolve this to a more sophisticated pattern, that will severely limit our possibilities for the next steps. Once we move on to dynamic workloads, we most definitely need to drop the "last inserted" policy since the last inserted task will most likely always be a fetched dependency of another task and it will be required immediately. Moving this off to another worker would be a horrible decision. I do not even want to start the discussion about what the best or even a good policy will be for this step, I just want to point out that this will be likely much more complicated and will require more flexibility.

My entire point is that we can already anticipate this insertion sort to not be sufficient. My proposal to include an explicit sort here was one possibility to make the rebalance modular and evolvable, not necessarily scalable. IMHO, your proposal suggests a highly scalable algorithm which is less modular and difficult to evolve. I believe we are looking for something in between.

File: distributed/utils_test.py [hunk 9]
```diff
@@ -1567,3 +1569,32 @@ def transition(self, key, start, finish, **kwargs):
...
+            finally:
+                importlib.reload(scheduler)
+
+        if scheduler.COMPILED:
+            return pytest.mark.skip(
+                "Monkey-patching module variables doesn't work with Cython"
+            )(wrapper)
+        return wrapper
+
+    return decorator
```
Comment (crusaderky) [inline distributed/utils_test.py:1600]: @jakirkham I could not make this work with Cython; could you take a look at it?
scheduler.py caches a bunch of dask.config keys for speed as global variables; I could not find a way to tamper with them when the module is cythonized.

- the design above does nothing with cython, in other words importlib.reload() seems to have no effect
- I also tried to monkey-patch the global variables themselves, but I get an AttributeError when Cython is enabled. I already tried changing scheduler.py to ``declare(..., visibility="public")`` but it does not seem to have an effect.

I guess a possible solution would be to move the dask.config.get calls from the module definition to ``Scheduler.__init__``?

**Comment by crusaderky** [comment]
Unit tests are done; rerunning them all once more now to prove stability.
Recommendations re. malloc_trim are to be reworked into memory.rst in the scope of a future PR. I have yet to go through proper tests on macosx in that regard.

Ready for final review.

**Comment by crusaderky** [comment]
All tests successful over 4 consecutive runs

**Comment by fjetter** [comment]
I left many comments but overall the code itself looks great. I also do like the clearly written tests since the expected behaviour is quite clear.

The two biggest issues I consider worth discussing are
* Global constants / config on import time
* Factor out _rebalance_find_msgs into dedicated functions to allow for scheduler/worker independent testing. I think this one is a bit of work but this could pay off nicely in the long run

File: distributed/scheduler.py [hunk 10]
```diff
@@ -161,14 +163,6 @@ def nogil(func):
 DEFAULT_DATA_SIZE = declare(
     Py_ssize_t, parse_bytes(dask.config.get("distributed.scheduler.default-data-size"))
 )
```
Comment (crusaderky) [inline distributed/scheduler.py:165]: could not move this as it is required by TaskState, and it would be way too expensive to reload it on every ``TaskState.__init__``

**Comment by crusaderky** [comment]
@fjetter 
> Global constants / config on import time

I redesigned this part. Now dask.config is read on ``Scheduler.__init__`` and I think it looks a lot better.

> Factor out _rebalance_find_msgs into dedicated functions to allow for scheduler/worker independent testing. I think this one is a bit of work but this could pay off nicely in the long run

While it is certainly possible to move _rebalance_find_msgs to a top-level function, it would require in input a SchedulerState object and a list of WorkerState objects, neither of which are easy - and most importantly, *robust* - to craft synthetically in a unit test. I'd much rather keep the tests as a black-box design because otherwise it's extremely easy that the real-life Scheduler and WorkerState objects will deviate from those crafted for the tests.

**Comment by fjetter** [comment]
> While it is certainly possible to move _rebalance_find_msgs to a top-level function, it would require in input a SchedulerState object and a list of WorkerState objects, neither of which are easy - and most importantly, robust - to craft synthetically in a unit test. 

We wouldn't need to pass the entire scheduler state in there, would we? I agree this would be nuts. From what I can see is that we're using four constants and the log_event method of the `SchedulerState` object. My point is that these constants can be a function input and we could simply not log this particular event. The event we're logging here is anyhow the "empty/nothing to do" event. I don't see a reason why this should be part of the core algorithm.

Passing in a list of `WorkerState` objects is exactly what I am proposing here and I believe it is more robust since we're not relying on futures, servers, comms, heartbeats, actual memory measurements or any timing relevant pieces. The tests would not even need to be sync/async since they would only test the actual algorithm.
Most attributes of the `WorkerState` object are irrelevant and we can run with the default values. The only actual coupling is introduced by the attributes we're using in the algorithm anyhow. Below an example how the test could look like

```python
def test_rebalance_simple():
    A = WorkerState(name="A", memory_limit="1GB")
    B = WorkerState(name="B", memory_limit="1GB")

    def put_task_on_worker(ws, key, nbytes):
        ws._has_what[key] = None
        ws._nbytes += nbytes

    for ix in range(4):
        put_task_on_worker(A, f"k{ix}", 200)

    result = _rebalance_find_msgs(workers=[A, B])

    expected = [
        (A, B, "k0"),
        (A, B, "k1"),
    ]

    assert result == expected
```

**Review by fjetter — COMMENTED** [review]
Ooops, had a few commends in 'pending', sorry

**Comment by crusaderky** [comment]
https://github.com/dask/distributed/pull/4853 breaks this. Temporarily reverting the other PR to prove that tests pass.

File: distributed/distributed-schema.yaml [hunk 11]
```diff
@@ -404,6 +404,49 @@ properties:
...
+                      receiving data from other workers. Ignored when max_memory is not
+                      set.
+
+                  recipient-max:
+                    type: number
+                    minimum: 0
+                    maximum: 1
+                    description: >-
+                      Fraction of worker process memory at which we start potentially
+                      transferring data to other workers.
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/distributed-schema.yaml:436]: yes. fixed

File: distributed/tests/test_client.py [hunk 12]
```diff
@@ -2891,92 +2891,119 @@ def __reduce__(self):
...
 
-def test_rebalance_sync(c, s, a, b):
-    futures = c.map(inc, range(10), workers=[a["address"]])
-    c.rebalance(futures)
+def test_rebalance_sync():
+    # can't use the 'c' fixture because we need workers to run in a separate process
+    with Client(n_workers=2, memory_limit="1 GiB") as c:
+        s = c.cluster.scheduler
+        a, b = [ws.address for ws in s.workers.values()]
+        futures = c.map(lambda _: "x" * (2 ** 29 // 10), range(10), workers=[a])
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/tests/test_client.py:2969]: fixed

File: distributed/tests/test_client.py [hunk 13]
```diff
@@ -2891,92 +2891,119 @@ def __reduce__(self):
...
+        assert 3 <= ndata[a] <= 7
+        assert 3 <= ndata[b] <= 7
 
 
 @gen_cluster(client=True)
 async def test_rebalance_unprepared(c, s, a, b):
+    """Client.rebalance() internally waits for unfinished futures"""
     futures = c.map(slowinc, range(10), delay=0.05, workers=a.address)
     await asyncio.sleep(0.1)
     await c.rebalance(futures)
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/tests/test_client.py:2988]: Added clarification in comments.
If a future errs, the whole rebalance will fall over with missing-data (see Client._rebalance). This has not changed from main. We could have a discussion as to if this behaviour is actually desirable, but it's out of scope of this PR.

File: distributed/tests/test_scheduler.py [hunk 14]
```diff
@@ -2496,3 +2497,251 @@ async def test_close_scheduler__close_workers_Nanny(s, a, b):
...
+    assert out == {"status": "missing-data", "keys": ["notexist"]}
+    s.validate_state()
+
+
+@gen_cluster(client=True)
+async def test_rebalance_missing_data2(c, s, a, b):
+    """keys exist but belong to unfinished futures"""
+    futures = c.map(slowinc, range(10), delay=0.05, workers=a.address)
+    await asyncio.sleep(0.1)
+    out = await s.rebalance(keys=[f.key for f in futures])
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/tests/test_scheduler.py:2599]: The tests in test_client.py test specifically the wrapper in client.py. Previously, all tests for rebalance were exclusively in test_client.py. However, very soon rebalance will be chiefly invoked internally by the scheduler, bypassing the client entirely. Hence, the tests should invoke Scheduler.rebalance and not rely on the client wrapper.

As to why Client.rebalance and Scheduler.rebalance behave differently: it would be non-trivial to move the logic upstream (Client uses Futures, Scheduler uses keys, for which I couldn't find a straightforward wait method?) and I just could not justify the effort. If in the future the Scheduler will call rebalance internally on unfinished keys, we will revisit this, but I just cannot see a use case in the plan that we laid down.

[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/tests/test_scheduler.py:2599]: Added clarification in the test that Client.rebalance and Scheduler.rebalance behave differently for this

File: distributed/tests/test_scheduler.py [hunk 15]
```diff
@@ -2496,3 +2497,251 @@ async def test_close_scheduler__close_workers_Nanny(s, a, b):
...
+    s.validate_state()
+
+
+@set_config_and_reload({"distributed.worker.memory.rebalance.measure": "managed"})
+@gen_cluster(client=True, Worker=Nanny, worker_kwargs={"memory_limit": "1000 MiB"})
+async def test_rebalance_managed_memory(c, s, *_):
+    a, b = s.workers
+    # Generate 100 buffers worth 400 MiB total on worker a. This sends its memory
+    # utilisation to exactly 40%, ignoring unmanaged, which is above the
+    # distributed.worker.memory.rebalance.sender-min threshold.
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/tests/test_scheduler.py:2638]: done

File: distributed/scheduler.py [hunk 16]
```diff
@@ -5507,144 +5527,363 @@ async def _delete_worker_data(self, worker_address, keys):
...
+            elif ws_memory < mean_memory - half_gap and ws_memory < recipient_max:
+                # This may send the worker above recipient_max (by design)
+                rec_bytes_max = ws_memory - mean_memory  # negative
+                rec_bytes_min = rec_bytes_max + half_gap  # negative
+                # See definition of recipients above
+                recipients.append((rec_bytes_max, rec_bytes_min, id(ws), ws))
+
+        # Fast exit in case no transfers are necessary or possible
+        if not senders or not recipients:
+            self.log_event(
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:5731]: refactoring the function is straightforward, but I don't agree on doing it to begin with. Refer to discussion above

**Comment by crusaderky** [comment]
> Passing in a list of WorkerState objects is exactly what I am proposing here and I believe it is more robust 

I strongly disagree. WorkerState objects are complex and in your code snippet you made very strong assumptions on their internal behaviour. If these internals changed in the future, I don't want a (potentially newbie) developer to have to hunt down unit tests that mock internal behaviour. What you're proposing has the big potential that, in a future PR completely unrelated to rebalance, rebalance will break without any of the unit tests noticing.

**Comment by mrocklin** [comment]
>> Passing in a list of WorkerState objects is exactly what I am proposing here and I believe it is more robust

> I strongly disagree. WorkerState objects are complex and in your code snippet you made very strong assumptions on their internal behaviour. If these internals changed in the future, I don't want a (potentially newbie) developer to have to hunt down unit tests that mock internal behaviour. What you're proposing has the big potential that, in a future PR completely unrelated to rebalance, rebalance will break without any of the unit tests noticing.

Just to jump in here with historical context.  Historically distributed had a lot of fine grained unit tests like this.  I've heard them called "white box" tests because they expose the internals, unlike "black box" tests.  In hindsight this ended up being a bad idea, especially as the scheduler changed internally.  Small changes to Scheduler internals required changing dozens of tests.  The tests became a liability and added inertia to development.  

At one point I spent a week rewriting white box tests so that we generally only touched user-level API.  This was *hard* because we needed to craft very specific situations where the desired behavior would arise.  In hindsight though this approach has provided a ton of value.  We're able to make large changes to internal scheduling state and the test suite remains valid.

File: distributed/tests/test_scheduler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/tests/test_scheduler.py:2606]: Thanks. The scheduler indeed does not have a way to wait for keys. The closes thing to this would be if we were to allow registration of "transition callbacks", such that if a key is transitioned into a given state, a callback is invoked. However, let's just not do that, the state machine is too complex as it is.

**Comment by fjetter** [comment]
> WorkerState objects are complex and in your code snippet you made very strong assumptions on their internal behaviour. 

As you said, this is a snippet I haven't put much thought into. The only relevant bit of this snippet is that the actual modifications are encapsulated into a dedicated function such that there is one place to change things. I didn't want to argue about the very specific way to do this until we reached a conclusion about whether this is something we want to have or not. Please do not reject this idea based on that snippet.

> What you're proposing has the big potential that, in a future PR completely unrelated to rebalance, rebalance will break without any of the unit tests noticing.

This assumes that there are no tests which cover the entire system and this is not something I propose. A bunch of the tests you wrote are still very important, exactly the way you wrote them, to know that everything works when put together. 
I'm proposing to implement tests covering the particular logic of the rebalancing algorithm as just that. The rebalancing algorithm is the solution to a problem "given X workers with keys Y and weights W and hardware measurements Z, calculate the necessary transitions such that the variance of weights per worker is minimal". This is formally not 100% correct but my point is that this is a purely mathematical problem and as such already a hard problem worth having enough test coverage.

The _how do we get the system into the proposed state_ is a hard problem by itself but less of mathematical nature. This problem is something where we need to deal with all the problems distributed systems introduce, like latencies, comm failures, dead nodes, race conditions, etc. Also a problem space where I think we should have enough edge case tests for non-happy paths.

Then there is the "plugging both pieces together" problem which doesn't require huge amounts of tests if both of the above subclasses are well tested.

> At one point I spent a week rewriting white box tests so that we generally only touched user-level API. This was hard because we needed to craft very specific situations where the desired behavior would arise. In hindsight though this approach has provided a ton of value. We're able to make large changes to internal scheduling state and the test suite remains valid.

I get that. That's the classical test hierarchy problem and both from my experience and from literature I believe that a healthy balance is good.
This flexibility may have given us the opportunity to rewrite a lot of code but we also dropped the ball on a few things we haven't had the chance to clean up, yet. Most of my time with the distributed code base, I was trying very hard to construct situation like you described but it took me much more than a week and we're still having issues actually reproducing some cases, in particular connected to the various race conditions and deadlocks associated to the worker state machine but not exclusively. 


The typical approach to this problem is to work with appropriate internal abstractions and (private) interfaces. If existing abstraction are not useful, let's create new ones or change existing ones. If the `WorkerState` is not a suitable interface, let's not use it. Maybe the `WorkerState` is indeed too complex and we should just provide _raw data_ (e.g. a dict per worker with the necessary measurements) to the algorithm. After all, we definitely don't want the algorithm part to actually mutate the state. Ideally, we wouldn't want to touch the algorithm at all if we were to change the `WorkerState` so there is no reason why it needs to be exposed to the actual objects. If we wanted to stick with the `WorkerState` for performance and complexity reasons we might want to look into whether the operations I proposed above are universal enough that it could become a method of the `WorkerState` object, making it less fragile to be used in tests since it will then be used all over the scheduler. Maybe an entirely different approach is better but I'm very certain we will find a robust interface if we want to. 
I'm actually more concerned with arguing that we want more of these thin internal interfaces than how they will particularly look like. In general I believe we should aim for having more of these since this allows for a less monolithic software architecture and allows us, if need be, to prepare a test system more easily for non-happy edge cases. This also helps with runtime and flakiness which is growing but that's only a secondary point.

**Comment by crusaderky** [comment]
Merged again from main after #4853 was fixed.
All tests pass.
All review comments with the exception of the refactoring have been addressed; @fjetter please let me know if there is anything else outstanding after my last few commits.

**Comment by fjetter** [comment]
Current test failures are (for reference)

https://github.com/dask/distributed/issues/4859 
https://github.com/dask/distributed/issues/4839
https://github.com/dask/distributed/issues/4862 (new but very likely unrelated)

---

I'll pass over the code once more but I don't expect something major to pop up. Let's have a chat about the possible refactoring this afternoon

**Review by fjetter — APPROVED** [review]
I had another look over the code and everything seems in order. Great job.

Regarding the refactoring, we couldn't reach a conclusion yet but I'm fine with merging as is. If we decide in favour of a refactoring, this will be additional work on top of this anyhow.

**Comment by mrocklin** [comment]
Woo!
